File: where_am_i/src/The_kidnapper_Algorithm.py
#!/usr/bin/env python3

import rospy
import tf2_ros
import tf_conversions
import geometry_msgs.msg
import numpy as np
import cv2
import matplotlib.pyplot as plt
from nav_msgs.msg import OccupancyGrid
import threading

class Matching:
    @staticmethod
    def publish_transform(center_x, center_y):
        br = tf2_ros.TransformBroadcaster()
        t = geometry_msgs.msg.TransformStamped()

        t.header.stamp = rospy.Time.now()
        t.header.frame_id = "map"
        t.child_frame_id = "base_link"

        # Convert pixel coordinates to map coordinates here
        map_x, map_y = Matching.pixel_to_map_conversion(center_x, center_y)

        # Set the position
        t.transform.translation.x = map_x
        t.transform.translation.y = map_y
        t.transform.translation.z = 0.0

        # Set the orientation
        q = tf_conversions.transformations.quaternion_from_euler(0, 0, 0)
        t.transform.rotation.x = q[0]
        t.transform.rotation.y = q[1]
        t.transform.rotation.z = q[2]
        t.transform.rotation.w = q[3]

        br.sendTransform(t)

    @staticmethod
    def pixel_to_map_conversion(pixel_x, pixel_y):
        return -pixel_x * 0.05, -pixel_y * 0.05  # pixel to real life measurement (resolution)
    
    @staticmethod
    def publish_initial_pose(x, y, theta):
        initial_pose_pub = rospy.Publisher('initialpose', geometry_msgs.msg.PoseWithCovarianceStamped, queue_size=10)
        rospy.sleep(1)  # Give time for the publisher to set up

        initial_pose = geometry_msgs.msg.PoseWithCovarianceStamped()
        initial_pose.header.stamp = rospy.Time.now()
        initial_pose.header.frame_id = "map"

        initial_pose.pose.pose.position.x = x
        initial_pose.pose.pose.position.y = y
        initial_pose.pose.pose.position.z = 0

        quaternion = tf_conversions.transformations.quaternion_from_euler(0, 0, theta)
        initial_pose.pose.pose.orientation.x = quaternion[0]
        initial_pose.pose.pose.orientation.y = quaternion[1]
        initial_pose.pose.pose.orientation.z = quaternion[2]
        initial_pose.pose.pose.orientation.w = quaternion[3]

        initial_pose_pub.publish(initial_pose)

    @staticmethod
    def match_template(image_path, template_path):
        bg = cv2.imread(image_path)
        bg = cv2.cvtColor(bg, cv2.COLOR_BGR2RGB)
        face = cv2.imread(template_path)
        face = cv2.cvtColor(face, cv2.COLOR_BGR2RGB)
        
        height, width, channels = face.shape
        methods = {"cv2.TM_CCOEFF": cv2.TM_CCOEFF}

        for method_name, method in methods.items():
            bg_copy = bg.copy()
            result = cv2.matchTemplate(bg_copy, face, method)
            min_val, max_val, min_loc, max_loc = cv2.minMaxLoc(result)

            if method in [cv2.TM_SQDIFF, cv2.TM_SQDIFF_NORMED]:
                top_left = min_loc
            else:
                top_left = max_loc

            center_x1 = top_left[0] + width // 2
            center_y1 = top_left[1] + height // 2
            center_x = abs((-center_x1))*0.05
            center_y = abs((-center_y1))*0.05
            bottom_right = (top_left[0] + width, top_left[1] + height)

            # Publish the transform and initial pose
            #publish_transform(center_x, center_y)
            Matching.publish_initial_pose(center_x, center_y, 0)  
            cv2.rectangle(bg_copy, top_left, bottom_right, 255, 10)

        # Plot the Images for debugging reason, you can delete it if you want
            plt.subplot(121)
            plt.imshow(result)
            plt.title("Result of Template Matching or photo matching")

            plt.subplot(122)
            plt.imshow(bg_copy)
            plt.title("Match Point")
            plt.show()


class Map:
    @staticmethod
    def map_value_to_color(value):
        if value == -1:
        # Unknown or unexplored
            return (128, 128, 128)  # Gray
        elif value <= 30:
        # Free space (0-30)
            return (255, 255, 255)  # White
        elif value <= 70:
            # Unknown space (30-70)
            return (0, 0, 255)  # Red
        else:
        # Occupied space (70-100)
            return (0, 255, 0)  # Green
    @staticmethod
    def callback(data):
    
        width = height = int(np.sqrt(len(data.data)))
        data_array = np.array(data.data, dtype=np.int8).reshape((height, width))

    
        color_image = np.array([Map.map_value_to_color(val) for val in data_array.flat], dtype=np.uint8).reshape((height, width, 3))

    
        cv2.imwrite("/home/alp/catkin_ws/src/nerdeyim/turtlebot3/where_am_i/src/local_colored.png", color_image)######### change it to your folder
        
        #rospy.sleep(5)

    @staticmethod
    def listener():
        #rospy.init_node('costmap_to_png_node', anonymous=True)
        rospy.Subscriber("/move_base/local_costmap/costmap", OccupancyGrid, Map.callback) #ensure that you have that cost map
        rospy.spin()

# Main loop
if __name__ == '__main__':
    rospy.init_node('Kidnapping_Preventation_Node', anonymous=True)
    listener_thread = threading.Thread(target=Map.listener)
    listener_thread.start()
    #to ensure that local costmap is written
    rospy.sleep(1)
    image_path = "/home/alp/catkin_ws/src/nerdeyim/turtlebot3/where_am_i/src/GLOBAL.png" ######### change it to your folder
    template_path = "/home/alp/catkin_ws/src/nerdeyim/turtlebot3/where_am_i/src/local_colored.png" ######### change it to your folder
    Localization = Matching()
    #Map.listener()

    try:
        Localization.match_template(image_path, template_path)
    except rospy.ROSInterruptException:
        pass
    finally:
        rospy.signal_shutdown('Task completed')  

File: where_am_i/src/the_kidnapper_algorithm.cpp
#include <ros/ros.h>
#include <tf2_ros/transform_listener.h>
#include <tf2/LinearMath/Quaternion.h>
#include <geometry_msgs/TransformStamped.h>
#include <opencv2/opencv.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <nav_msgs/OccupancyGrid.h>
#include <thread>
#include <tf2_ros/transform_broadcaster.h>
#include <geometry_msgs/PoseWithCovarianceStamped.h>
#include <iostream>
#include <vector>

class Matching {
public:
    static void publishTransform(double center_x, double center_y) {
        static tf2_ros::TransformBroadcaster br;
        geometry_msgs::TransformStamped t;

        t.header.stamp = ros::Time::now();
        t.header.frame_id = "map";
        t.child_frame_id = "base_link";

        // Convert pixel coordinates to map coordinates
        auto [map_x, map_y] = pixelToMapConversion(center_x, center_y);

        // Set the position
        t.transform.translation.x = map_x;
        t.transform.translation.y = map_y;
        t.transform.translation.z = 0.0;

        // Set the orientation
        tf2::Quaternion q;
        q.setRPY(0, 0, 0);
        t.transform.rotation.x = q.x();
        t.transform.rotation.y = q.y();
        t.transform.rotation.z = q.z();
        t.transform.rotation.w = q.w();

        br.sendTransform(t);
    }

    static std::pair<double, double> pixelToMapConversion(double pixel_x, double pixel_y) {
        return {-pixel_x * 0.05, -pixel_y * 0.05}; // pixel to real life measurement conversion
    }

    static void publishInitialPose(double x, double y, double theta) {
        ros::NodeHandle nh;
        ros::Publisher initial_pose_pub = nh.advertise<geometry_msgs::PoseWithCovarianceStamped>("initialpose", 10);
        ros::Duration(1).sleep(); // Give time for the publisher to set up

        geometry_msgs::PoseWithCovarianceStamped initial_pose;
        initial_pose.header.stamp = ros::Time::now();
        initial_pose.header.frame_id = "map";

        initial_pose.pose.pose.position.x = x;
        initial_pose.pose.pose.position.y = y;
        initial_pose.pose.pose.position.z = 0;

        tf2::Quaternion quaternion;
        quaternion.setRPY(0, 0, theta);
        initial_pose.pose.pose.orientation.x = quaternion.x();
        initial_pose.pose.pose.orientation.y = quaternion.y();
        initial_pose.pose.pose.orientation.z = quaternion.z();
        initial_pose.pose.pose.orientation.w = quaternion.w();

        initial_pose_pub.publish(initial_pose);
    }

    static void matchTemplate(const std::string& image_path, const std::string& template_path) {
        cv::Mat bg = cv::imread(image_path);
        cv::cvtColor(bg, bg, cv::COLOR_BGR2RGB);
        cv::Mat face = cv::imread(template_path);
        cv::cvtColor(face, face, cv::COLOR_BGR2RGB);

        int height = face.rows;
        int width = face.cols;

        cv::Mat result;
        cv::matchTemplate(bg, face, result, cv::TM_CCOEFF);

        double minVal, maxVal;
        cv::Point minLoc, maxLoc;
        cv::minMaxLoc(result, &minVal, &maxVal, &minLoc, &maxLoc);

        cv::Point top_left = maxLoc;
        double center_x1 = top_left.x + width / 2;
        double center_y1 = top_left.y + height / 2;
        double center_x = std::abs((-center_x1)) * 0.05;
        double center_y = std::abs((-center_y1)) * 0.05;
        cv::Point bottom_right = top_left + cv::Point(width, height);

        // Publish the transform and initial pose
        // publishTransform(center_x, center_y); // Uncomment if needed
        publishInitialPose(center_x, center_y, 0);
        cv::rectangle(bg, top_left, bottom_right, cv::Scalar(255, 0, 0), 10);

        // Display the images
        cv::imshow("Result of Template Matching", result);
        cv::imshow("Match Point", bg);
        cv::waitKey(0);
    }
};


class Map {
public:
    static cv::Scalar mapValueToColor(int value) {
        if (value == -1) {
            // Unknown or unexplored
            return cv::Scalar(128, 128, 128); // Gray
        } else if (value <= 30) {
            // Free space (0-30)
            return cv::Scalar(255, 255, 255); // White
        } else if (value <= 70) {
            // Unknown space (30-70)
            return cv::Scalar(0, 0, 255); // Red
        } else {
            // Occupied space (70-100)
            return cv::Scalar(0, 255, 0); // Green
        }
    }

    static void callback(const nav_msgs::OccupancyGrid::ConstPtr& data) {
        int width = data->info.width;
        int height = data->info.height;
        std::vector<int8_t> data_vector(data->data.begin(), data->data.end());

        cv::Mat color_image(height, width, CV_8UC3);
        for (int y = 0; y < height; ++y) {
            for (int x = 0; x < width; ++x) {
                int index = y * width + x;
                cv::Scalar color = mapValueToColor(data_vector[index]);
                color_image.at<cv::Vec3b>(cv::Point(x, y)) = cv::Vec3b(color[0], color[1], color[2]);
            }
        }

        cv::imwrite("/home/alp/catkin_ws/src/turtlebot3_kalman_filter/src/local_colored.png", color_image);
    }

    static void listener() {
        ros::NodeHandle nh;
        ros::Subscriber sub = nh.subscribe("/move_base/local_costmap/costmap", 1000, callback);
        ros::spin();
    }
};

int main(int argc, char **argv) {
    ros::init(argc, argv, "Kidnapping_Prevention_Node");
    std::thread listener_thread(&Map::listener);
    listener_thread.detach(); // Detach the thread so that it runs independently

    std::string image_path = "/home/alp/catkin_ws/src/turtlebot3_kalman_filter/src/alp.png";
    std::string template_path = "/home/alp/catkin_ws/src/turtlebot3_kalman_filter/src/local_colored.png";
    
    try {
        Matching::matchTemplate(image_path, template_path);
    } catch (const ros::Exception& e) {
        ROS_ERROR("ROS exception: %s", e.what());
    }
    ros::shutdown();
    ROS_INFO("Location Founded");
    

    return 0;
}
